File: leetcode/Merge Two Sorted Lists.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution:
    # @param two ListNodes
    # @return a ListNode
    def mergeTwoLists(self, l1, l2):
        if not l1 and not l2:
            return None
        if not l1 and l2:
            return l2
        if not l2 and l1:
            return l1
        p1, p2 = l1, l2
        head = ListNode(0)
        p = head
        while p1 or p2:
            if not p1:
                p.next = ListNode(p2.val)
                p2 = p2.next
            elif not p2:
                p.next = ListNode(p1.val)
                p1 = p1.next
            elif p1.val < p2.val:
                p.next = ListNode(p1.val)
                p1 = p1.next
            elif p2.val < p1.val:
                p.next = ListNode(p2.val)
                p2 = p2.next
            else:
                p.next = ListNode(p1.val)
                p1 = p1.next
            p = p.next
        return head.next

File: leetcode/cpp/Merge_Two_Sorted_Lists.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    ListNode* mergeTwoLists(ListNode* l1, ListNode* l2) {
        if (NULL == l1 && NULL == l2)
            return NULL;
            
        ListNode first(0);
        ListNode *dummy = &first;
        
        ListNode *p1 = l1, *p2 = l2;
        while (p1 && p2) {
            if (p1->val > p2->val) {
                dummy->next = new ListNode(p2->val);
                p2 = p2->next;
            }
            else {
                dummy->next = new ListNode(p1->val);
                p1 = p1->next;
            }
            dummy = dummy->next;
        }
        
        if (NULL == p1) {
            copyList(dummy, p2);
        }
        else {
            copyList(dummy, p1);
        }
        return first.next;
    }
    
    void copyList(ListNode *des, ListNode *src) {
        while (src) {
            des->next = new ListNode(src->val);
            des = des->next;
            src = src->next;
        }
    }
};